src/Common/src/Interop/Linux/System.Net.Security.Native/Interop.Ntlm.cs

```diff
@@ -0,0 +1,124 @@
+        internal static extern int HeimNtlmCalculateResponse(
+            bool isLM,
+            SafeNtlmBufferHandle key,
+            SafeNtlmType2Handle type2Handle,
+            string username,
+            string target,
+            byte[] baseSessionKey,
+            int baseSessionKeyLen,
+            out SafeNtlmBufferHandle answer,
+            out int ansLength);
+
+        [DllImport(Interop.Libraries.SecurityNative, EntryPoint="NetSecurity_CreateType3Message", CharSet = CharSet.Ansi)]
+        internal static extern int CreateType3Message(
+            SafeNtlmBufferHandle key,
+            SafeNtlmType2Handle type2Handle,
+            string username,
+            string domain,
+            uint flags,
+            SafeNtlmBufferHandle lmResponse,
+            SafeNtlmBufferHandle ntlmResponse,
+            byte [] baseSessionKey,
+            int baseSessionKeyLen,
+            out SafeNtlmBufferHandle sessionKey,
+            out int sessionKeyLen,
+            out SafeNtlmBufferHandle data,
+            out int dataLen
+            );
+
+        internal partial class NtlmFlags
+        {
+            internal const uint NTLMSSP_NEGOTIATE_UNICODE = 0x1;
+            internal const uint NTLMSSP_REQUEST_TARGET = 0x4;
+            internal const uint NTLMSSP_NEGOTIATE_SIGN = 0x10;
+            internal const uint NTLMSSP_NEGOTIATE_SEAL = 0x20;
+            internal const uint NTLMSSP_NEGOTIATE_NTLM = 0x200;
+            internal const uint NTLMSSP_NEGOTIATE_ALWAYS_SIGN = 0x8000;
+            internal const uint NTLMSSP_NEGOTIATE_EXTENDED_SESSIONSECURITY = 0x80000;
+            internal const uint NTLMSSP_NEGOTIATE_128 = 0x20000000;
+            internal const uint NTLMSSP_NEGOTIATE_KEY_EXCH = 0x40000000;
+        }
+
+        internal static byte[] EVPDigest(byte[] key, byte[] input, int inputlen, out uint outputlen)
+        {
+            //reference doc: http://msdn.microsoft.com/en-us/library/cc236700.aspx
+            byte[] output = new byte[Interop.Crypto.EVP_MAX_MD_SIZE];
+            outputlen = 0;
+            using (SafeEvpMdCtxHandle ctx = Interop.Crypto.EvpMdCtxCreate(Interop.Crypto.EvpMd5()))
+            unsafe
+            {
+                fixed (byte *keyPtr = key, inPtr = input, outPtr = output)
+                {
+                    Check(Interop.Crypto.EvpDigestUpdate(ctx, keyPtr, key.Length));
+                    Check(Interop.Crypto.EvpDigestUpdate(ctx, inPtr, inputlen));
+                    Check(Interop.Crypto.EvpDigestFinalEx(ctx, outPtr, ref outputlen));
+                }
+            }
+            return output;
+        }
+
+        internal static unsafe byte[] HMACDigest(byte* key, int keylen, byte* input, int inputlen, byte* prefix, int prefixlen, out int hashLength)
+        {
+            hashLength = 0;
+            byte[] output = new byte[Interop.Crypto.EVP_MAX_MD_SIZE];
+            using (SafeHmacCtxHandle ctx = Interop.Crypto.HmacCreate(key, keylen, Interop.Crypto.EvpMd5()))
```

Comment:
And this is just an incremental HMAC... Again, why not just use the IncrementalHash class?